pkg/kbuffer/endpoint_observer.go

```diff
@@ -0,0 +1,104 @@
+package kbuffer
```

Comment:
@josephburnett we're saying the same thing I think. In fact, @vvraskin already used an Informer here.

Reply:
The funny thing is that there is a new call to `Watch` for each first request to a revision.  An each time it adds a handler.  I don't really see where we remove these handlers.

What about this instead:

1. The throttler keeps an up-to-date view of Endpoints through the informer.
2. The throttler also keeps a set of breakers, one for each Endpoint.
3. As endpoints are created / updated, it adjusted the `maxConcurrency` to equal the number of active endpoints (initially 0).
4. When the `DedupingActivator` wants to release pending requests, it get the breaker for the endpoint from the throttle and makes all the calls through that.

Reply:
@josephburnett in `2` you probably meant one Broker for one Revision, not Endpoint?

Reply:
Thinking about it - actually it would make more sense to have a single broker for each Endpoint. And use Throttler just to round robin across the brokers?
Otherwise (single broker - multiple endpoints) we need to add a dance to dynamically resize its channels - which won't be easy I think.

Reply:
I'm not 100% confident about a one breaker per endpoint approach. When routing traffic from the activator to the revision, we don't generally send them to a specific pod (today) but send them to the revision via istio, which does the round-robin. In a sense, this *can* work, if we try each breaker and then if we don't succeed to get through, place the request on a queue to wait.

What we need is a semaphore that's sized according to the amount of endpoints * containerConcurrency per Revision basically. Having N Semaphores (our breakers) of a fixed length could be good enough in our case though, consistency is not a hard requirement for the task at hand I think.

Reply:
In the case with semaphores - we cannot use the breakers as they are today - their queue size cannot be adjusted dynamically that easily. We would need to implement a dynamic semaphore or reuse something like this - https://github.com/marusama/semaphore
Thus I'm more in favour of one breaker per endpoint approach. As you said: we can first top up all available brokers (endpoints) and then spray across them randomly.